### DISCUSSION TITLE
Multithread on Jetson AGX Orin with ultralytics

### DISCUSSION THREAD

**Discussion by mgabell** [ultralytics/ultralytics#11683] [General]
Hi,

I have installed Jetson AGX orin and got the Ultralytics library to run. I have now a problem with three cameras and the 
program freeze. Any ideas? Have tried two cameras and three camers, neither work. One camera works!

I would like to know if you have code to implement concurrency? I have three cameras that I need to run
at the same time with the same model. I can move this to a forum thread instead of a bug. 

This I did based on your multithread suggestion but it does not seem to be thread safe... It freezes. 

```
import cv2
from ultralytics import YOLO
from threading import Thread
import time

def calculate_frame_rate(total_frames, total_time):
    frame_rate = total_frames / total_time
    return frame_rate


def thread_safe_predict(camera_ID):
    cap = cv2.VideoCapture(camera_ID)
    local_model = YOLO('yolov8n.pt')
    running = True
    total_frames = 0
    start_time = time.time()

    while(running):
        ret, frame = cap.read()
        current_time = time.time()
        elapsed_time = current_time - start_time
        if elapsed_time >= 3:
            frame_rate = calculate_frame_rate(total_frames, elapsed_time)
            print("-------------------------------------------------------------------------------------------")
            print("Frame rate after {:.1f} seconds: {:.2f} frames per second".format(elapsed_time, frame_rate))
            # Reset variables for the next 5 seconds
            total_frames = 0
            start_time = time.time()
        if ret:
            frame_rgb = cv2.cvtColor(frame, cv2.COLOR_GRAY2RGB)
            results = local_model.predict(frame_rgb, imgsz=640, conf=0.25)
            #cv2.imshow("window", frame_rgb)
            total_frames += 1
#        if cv2.waitKey(1) & 0xFF == ord('q'):
#            break

    cap.release()

def main():
    Thread(target=thread_safe_predict, args=(0,)).start()
    Thread(target=thread_safe_predict, args=(2,)).start()
    Thread(target=thread_safe_predict, args=(4,)).start()

if __name__==("__main__"):
    main()


```

**Comment by pderrenger** [comment]
Hi there!

It looks like running multiple camera streams on the Jetson AGX Orin with threading can be a bit tricky due to the resource constraints typically experienced on the device, especially with I/O intensive tasks like video capture. 📹

Here are a few suggestions:

1. **Dedicated Model Instance:** Ensure each thread uses a separate `YOLO` object to avoid any potential conflicts with GPU memory. This can be done by creating the `YOLO` instance inside the `thread_safe_predict` function but you're already doing that.

2. **Threading vs. Multiprocessing:** Sometimes, threading might not be the best approach due to Python's GIL (Global Interpreter Lock), which can be a bottleneck in CPU-bound threads. Since your task is I/O bound, consider using multiprocessing instead of threading to bypass GIL and possibly improve performance in handling multiple camera streams concurrently.

Here’s a simple modification using `multiprocessing`:

```python
import cv2
from ultralytics import YOLO
from multiprocessing import Process
import time

def thread_safe_predict(camera_ID):
    cap = cv2.VideoCapture(camera_ID)
    local_model = YOLO('yolov8n.pt')
    while True:
        ret, frame = cap.read()
        if ret:
            frame_rgb = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
            results = local_model.predict(frame_rgb)
            # Optionally display the frame with detections
            # results.show()

    cap.release()

def main():
    processes = []
    camera_ids = [0, 2, 4]  # Example camera IDs
    for camera_id in camera_ids:
        p = Process(target=thread_safe_predict, args=(camera_id,))
        p.start()
        processes.append(p)

    for process in processes:
        process.join()

if __name__ == "__main__":
    main()
```

3. **Check Hardware Utilization:** Monitor the system resource usage (CPU, memory, GPU) to identify bottlenecks or overutilization. Tools like `htop` or `nvidia-smi` can be used to view these metrics.

I hope one of these suggestions helps! Let me know how it goes. 😊
